core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2690,7 +2694,14 @@ int complexity() {
     }
 
     @Override public RelDataType getType() {
-      return rexBuilder.typeFactory.leastRestrictive(Util.distinctList(types));
+      List<RelDataType> distinctTypes = Util.distinctList(this.types);
+      assert distinctTypes.size() >= 1 : "at least one type expected";
+      if (distinctTypes.size() == 1) {
+        return distinctTypes.get(0);
```

Comment:
Not necessarily subject of this PR but couldn't this logic be inside `leastRestrictive`? If there is only one type then it is also the least restrictive so putting it inside we don't break anything in terms of API. Moreover, if we want to force users to pass more than one type then it would be better to check this via regular exceptions (e.g., `IllegalArgumentException`) instead of assertions that can even be disabled.

Reply:
It probably makes sense to move this `if size=1` check to the `leastRestrictive` method.

However, I do not know if `null` result is permissible here or if it is a catastrophic failure, so I added an assertion, except I put the relevant clarification to the message.